libbpf-tools/trace_helpers.c

```diff
@@ -159,6 +166,103 @@ const struct ksym *ksyms__get_symbol(const struct ksyms *ksyms,
 	return NULL;
 }
 
+struct partitions {
+	struct partition *items;
+	int sz;
+};
+
+static int partitions__add_partition(struct partitions *partitions,
+				     const char *name, unsigned int dev)
+{
+	struct partition *partition;
+	void *tmp;
+
+	tmp = realloc(partitions->items, (partitions->sz + 1) *
+		sizeof(*partitions->items));
+	if (!tmp)
+		return -1;
+	partitions->items = tmp;
+	partition = &partitions->items[partitions->sz];
+	partition->name = strdup(name);
+	partition->dev = dev;
+	partitions->sz++;
+
+	return 0;
+}
+
+struct partitions *partitions__load(void)
+{
+	char part_name[DISK_NAME_LEN];
+	unsigned int devmaj, devmin;
+	unsigned long long nop;
+	struct partitions *partitions;
+	char buf[64];
+	FILE *f;
+
+	f = fopen("/proc/partitions", "r");
+	if (!f)
+		return NULL;
+
+	partitions = calloc(1, sizeof(*partitions));
+	if (!partitions)
+		goto err_out;
+
+	while (fgets(buf, sizeof(buf), f) != NULL) {
+		/* skip heading */
+		if (buf[0] != ' ' || buf[0] == '\n')
+			continue;
+		if (sscanf(buf, "%u %u %llu %s", &devmaj, &devmin, &nop,
+				part_name) != 4)
+			goto err_out;
+		if (partitions__add_partition(partitions, part_name,
+						MKDEV(devmaj, devmin)))
+			goto err_out;
+	}
+
+	fclose(f);
+	return partitions;
+
+err_out:
+	partitions__free(partitions);
+	fclose(f);
+	return NULL;
+}
+
+void partitions__free(struct partitions *partitions)
+{
+	if (!partitions)
+		return;
+
+	free(partitions->items);
```

Comment:
you need to iterate over partitions->items and free ->name as well

Reply:
Fixed, so terrible, I'm going to face the wall and think ...